lib/stm32/common/st_usbfs_core.c

```diff
@@ -97,27 +109,51 @@ void st_usbfs_ep_setup(usbd_device *dev, uint8_t addr, uint8_t type,
 	USB_SET_EP_TYPE(addr, typelookup[type]);
 
 	if (dir || (addr == 0)) {
-		USB_SET_EP_TX_ADDR(addr, dev->pm_top);
+		if (type == USB_ENDPOINT_ATTR_ISOCHRONOUS) {
+			/* isochronous endpoints always use double buffering
+			 * and are unidirectional */
+			USB_SET_EP_RX_STAT(addr, USB_EP_RX_STAT_DISABLED);
+			USB_SET_EP_TX_0_ADDR(addr, dev->pm_top);
+			USB_SET_EP_TX_0_COUNT(addr, max_size);
+			dev->pm_top += max_size;
+			USB_SET_EP_TX_1_ADDR(addr, dev->pm_top);
+			USB_SET_EP_TX_1_COUNT(addr, max_size);
+		} else {
+			USB_SET_EP_TX_ADDR(addr, dev->pm_top);
+			USB_SET_EP_TX_COUNT(addr, max_size);
+		}
+		dev->pm_top += max_size;
 		if (callback) {
 			dev->user_callback_ctr[addr][USB_TRANSACTION_IN] =
 			    (void *)callback;
 		}
 		USB_CLR_EP_TX_DTOG(addr);
-		USB_SET_EP_TX_STAT(addr, USB_EP_TX_STAT_NAK);
-		dev->pm_top += max_size;
+		if (type == USB_ENDPOINT_ATTR_ISOCHRONOUS)
+			USB_SET_EP_TX_STAT(addr, USB_EP_TX_STAT_VALID); // or DISABLED?
```

Comment:
what's the "or disabled" for? still unsure? please be sure? :)

Reply:
Further, always, always, always use properly braced {} statements.

Reply:
Oops forgot to remove that. It works like it is, with my testing.